request-cert/main.go

```diff
@@ -0,0 +1,190 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Marc berhault (marc@cockroachlabs.com)
+
+package main
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	certificateType = flag.String("type", "", "certificate type: node or client")
+
+	addresses = flag.String("addresses", "", "comma-separated list of DNS names and IP addresses for node certificate")
+	user      = flag.String("user", "", "username for client certificate")
+
+	certsDir        = flag.String("certs-dir", "cockroach-certs", "certs directory")
+	keySize         = flag.Int("key-size", 2048, "RSA key size in bits")
+	symlinkCASource = flag.String("symlink-ca-from", "", "if non-empty, create <certs-dir>/ca.crt linking to this file")
+)
+
+func main() {
+	flag.Parse()
+	flag.Lookup("logtostderr").Value.Set("true")
+
+	// Check certificate type.
+	var template *x509.CertificateRequest
+	var filePrefix string
+	var wantServerAuth bool
+
+	switch *certificateType {
+	case "node":
+		if len(*addresses) == 0 {
+			log.Fatal("node certificate requested, but --addresses is empty")
+		}
+		template = serverCSR(strings.Split(*addresses, ","))
+		filePrefix = "node"
+		wantServerAuth = true
+	case "client":
+		if len(*user) == 0 {
+			log.Fatal("client certificate requested, but --user is empty")
+		}
+		template = clientCSR(*user)
+		filePrefix = "client." + *user
+	default:
+		log.Fatalf("unknown certificate type requested: --type=%q. Valid types are \"node\", \"client\"", *certificateType)
+	}
+
+	// Generate a new private key.
+	privateKey, err := rsa.GenerateKey(rand.Reader, *keySize)
+	if err != nil {
+		log.Fatalf("error generating RSA key pair: %v", err)
+	}
+
+	// Generate CSR. The helper splits addresses into IP or DNS names.
+	pemCSR, err := templateToCSRBytes(template, privateKey)
+	if err != nil {
+		log.Fatalf("error generating CSR: %v", err)
+	}
+
+	// Try to build a descriptive name for the CSR, that's all the admin sees.
+	csrName := filePrefix
+	if hostname, err := os.Hostname(); err == nil && len(hostname) != 0 {
+		csrName += "-" + hostname
+	}
+
+	pemCert, err := kubernetesSigner(csrName, pemCSR, wantServerAuth)
+	if err != nil {
+		log.Fatalf("CSR signing failed: %v", err)
+	}
+
+	if err := writeFiles(filePrefix, privateKey, pemCert); err != nil {
+		log.Fatalf("failed to write files: %v", err)
+	}
+}
+
+// serverCSR generates a certificate signing request for a server certificate and returns it.
+// Takes in the list of hosts/ip addresses this certificate applies to.
+func serverCSR(hosts []string) *x509.CertificateRequest {
+	csr := &x509.CertificateRequest{
+		Subject: pkix.Name{
+			Organization: []string{"Cockroach"},
```

Comment:
This has been around pretty much since we've started doing certificates.
We don't currently look at, let alone enforce, the `Subject` field other than `CommonName`. I have plans to start doing so but until then, I would rather leave things identical between `cockroachdb/cockroach/security/` and this code.